geopandas/tools/sjoin.py

```diff
@@ -186,6 +168,429 @@ def find_contains(a1, a2):
+    left_df, right_df, left_index_name, right_index_name, index_left, index_right \
+     = _rename_indexes(left_df,right_df,lsuffix,rsuffix)
+
+    # ------------------------ COMPUTE SPATIAL JOIN ----------------------------
+
+    # get spatial index
+    tree_idx = right_df.sindex
+
+    # validate max_search_neighbors and search_radius params
+    if max_search_neighbors is not None:
+        assert max_search_neighbors >= 1 and isinstance(max_search_neighbors, int), ""
+        "max_search_neighbors must be an integer >= 1"
+    if search_radius is not None:
+        assert search_radius >= 0, "search_radius must be positive"
+
+    def query_index(geo_in_l):
+        """
+        Queries the spatial index to filter results.
+        """
+        # restric by radius
+        if search_radius is not None:
+            if geo_in_l.geom_type == 'Point' and search_radius == 0:
+                # applying .buffer(0) to a point returns an empty geometry
+                geo_search = geo_in_l.coords[0]
+            else: # polygon or buffer > 0
+                geo_search = geo_in_l.buffer(search_radius).bounds
+            in_radius = list(tree_idx.intersection(geo_search))
+        else:
+            in_radius = right_df.index
+        # find neighbors
+        if max_search_neighbors is not None:
+            neighbors = list(tree_idx.nearest(
+                geo_in_l.bounds, num_results=max_search_neighbors 
+                ))
+        else:
+            neighbors = right_df.index
+        return (idx for idx in neighbors if idx in in_radius)
+
+    # check each set of matches for actual distance
+    l_idx = []
+    r_idx = []
+    distances = []
+    for ind_in_left, geo_in_l in enumerate(left_df.geometry):
+        if not geo_in_l or geo_in_l.is_empty:
+            # see https://github.com/Toblerity/Shapely/issues/799
+            continue
+        possible_nearest = query_index(geo_in_l)
+        min_dist = np.inf  # initialize
+        min_ind = [] # initialize
+        for ind_in_right in possible_nearest:
+            geo_in_r = right_df.geometry.iloc[ind_in_right]
+            if not geo_in_r or geo_in_r.is_empty:
+                # see https://github.com/Toblerity/Shapely/issues/799
+                continue
+            dist = geo_in_l.distance(geo_in_r)
+            if dist < min_dist and (search_radius is None or dist <= search_radius):
+                # new closest
+                min_dist = dist  # reset
+                min_ind = [ind_in_right]  # re-initialize
+            elif dist == min_dist:
+                # matching closest, extend list of closest
+                min_ind.append(ind_in_right)
+        # extend final results
+        if min_ind: # possible_nearest may have been empty
+            r_idx.extend(min_ind)
+            l_idx.extend([ind_in_left] * len(min_ind))
+            if nearest_distances:  # avoid memory use if unwarrented
+                distances.extend([min_dist] * len(min_ind))
```

Comment:
This is the algorithm part.

I tried implementing this making more use of `numpy` vectorized methods. Maybe it could be done better, but I actually saw it perform _worse_ than the implementation here (and I'm sure it's using a lot more memory).

I also tested the effects of prepared geometries (I just deleted `.apply(lambda x: prepared.prep(x))` in `sjoin`).
With prepared geometries, `sjoin` is ~3x faster than `sjoin_nearest` with `search_radius=0` and `max_search_neighbors=None`. Without prepared geometries, `sjoin_nearest` is ~ 30% faster.

I think this is confirmation that we won't be able to improve the performance unless we can get `shapely` to implement `distance` for prepared geometries or we move to a faster underlying geometry data structure.